### PR TITLE
[FEAT] databricks repair run with reason match and appropriate new settings

### PR DESCRIPTION
**Pull request by gaurav7261**
[FEAT] databricks repair run with reason match and appropriate new settings in `DatabricksRunNowOperator`
This is useful in scenario like where we get databricks exception like `AWS_INSUFFICIENT_INSTANCE_CAPACITY_FAILURE (CLIENT_ERROR).`,  `AWS_MAX_SPOT_INSTANCE_COUNT_EXCEEDED_FAILURE ` where user want to repair run with let's say different zone of region or let's say with 100% ondemand  

**^ Add meaningful description above**
Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/contributing-docs/05_pull_requests.rst#pull-request-guidelines)** for more information.
In case of fundamental code changes, an Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvement+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in a newsfragment file, named `{pr_number}.significant.rst` or `{issue_number}.significant.rst`, in [newsfragments](https://github.com/apache/airflow/tree/main/newsfragments).

### CODE REVIEW

File: airflow/providers/databricks/operators/databricks.py [hunk 1]
```diff
@@ -107,6 +107,22 @@ def _handle_databricks_operator_execution(operator, hook, log, context) -> None:
...
+                        None,
+                    )
+
+                    if repair_reason is not None and operator.databricks_repair_reason_new_settings:
+                        new_settings_json = normalise_json_content(
+                            operator.databricks_repair_reason_new_settings[repair_reason]
+                        )
+                        log.warning("Repairing the run with new_settings json: %s", new_settings_json)
+                        hook.update_job(job_id=operator.json["job_id"], json=new_settings_json)
+
```
Comment (shahar1) [inline airflow/providers/databricks/operators/databricks.py]: Could you please try adding tests for this part?

File: airflow/providers/databricks/operators/databricks.py [hunk 2]
```diff
@@ -872,7 +896,23 @@ def execute_complete(self, context: Context, event: dict[str, Any] | None = None
...
+                        for reason in (self.databricks_repair_reason_new_settings or {}).keys()
+                        if reason in run_state.state_message
+                    ),
+                    None,
+                )
+                if repair_reason is not None and self.databricks_repair_reason_new_settings:
+                    new_settings_json = normalise_json_content(
+                        self.databricks_repair_reason_new_settings[repair_reason]
+                    )
+                    self._hook.update_job(job_id=str(job_id), json=new_settings_json)
```
Comment (shahar1) [inline airflow/providers/databricks/operators/databricks.py]: 1. Same comment regarding adding tests.
2. Looks repetitive - maybe it's worth extracting into a util function/static method and reuse it?

File: airflow/providers/databricks/operators/databricks.py [hunk 3]
```diff
@@ -107,6 +107,22 @@ def _handle_databricks_operator_execution(operator, hook, log, context) -> None:
                         error_message,
                     )
 
+                    repair_reason = next(
+                        (
+                            reason
+                            for reason in (operator.databricks_repair_reason_new_settings or {}).keys()
```
Comment (Lee-W) [inline airflow/providers/databricks/operators/databricks.py]:
```suggestion
                            for reason in operator.databricks_repair_reason_new_settings
```

File: airflow/providers/databricks/operators/databricks.py (hunk 2 above)
Comment (Lee-W) [reply airflow/providers/databricks/operators/databricks.py]: +1

Comment (gaurav7261) [reply airflow/providers/databricks/operators/databricks.py]: sure working on it

**Comment by gaurav7261** [comment]
@Lee-W @shahar1 can you please review now

**Review by shahar1 — APPROVED** [review]
LGTM

File: airflow/providers/databricks/operators/databricks.py [hunk 4]
```diff
@@ -123,6 +124,30 @@ def _handle_databricks_operator_execution(operator, hook, log, context) -> None:
...
+    """
+    repair_reason = next(
+        (
+            reason
+            for reason in operator.databricks_repair_reason_new_settings
+            if reason in run_state.state_message
+        ),
+        None,
+    )
+    if repair_reason is not None and operator.databricks_repair_reason_new_settings:
```
Comment (Lee-W) [inline airflow/providers/databricks/operators/databricks.py]: Would like to check whether we need the later `operator.databricks_repair_reason_new_settings` check. When will `repair_reason` not be none and `operator.databricks_repair_reason_new_settings` be empty at the same time?

File: airflow/providers/databricks/operators/databricks.py [hunk 5]
```diff
@@ -123,6 +124,30 @@ def _handle_databricks_operator_execution(operator, hook, log, context) -> None:
     log.info("View run status, Spark UI, and logs at %s", run_page_url)
 
 
+def update_job_for_repair(operator, hook, job_id, run_state) -> None:
```
Comment (Lee-W) [inline airflow/providers/databricks/operators/databricks.py]: would be great if we can improve type annotaiton

**Review by Lee-W — COMMENTED** [review]
no huge issue, just a few nitpicks

File: airflow/providers/databricks/operators/databricks.py (hunk 4 above)
Comment (gaurav7261) [reply airflow/providers/databricks/operators/databricks.py]: yes we don't need it, fixing

File: airflow/providers/databricks/operators/databricks.py (hunk 5 above)
Comment (gaurav7261) [reply airflow/providers/databricks/operators/databricks.py]: working on it

**Comment by gaurav7261** [comment]
@Lee-W require changes done, also added assert_called_with check for `update_job` method as well.

**Comment by gaurav7261** [comment]
@Lee-W  should we have one more configurable like user want to repair only in case if `databricks_repair_reason_new_settings` dict key contains the reason, otherwise don't repair at all(as it is a legit error(like some job issue) and most probably will not get resolve even with one more repair run), I want to ask what you think for this ? and how can we enable that as well(have one more argument or something in your mind) ?

**Comment by Lee-W** [comment]
> @Lee-W should we have one more configurable like user want to repair only in case if `databricks_repair_reason_new_settings` dict key contains the reason, otherwise don't repair at all(as it is a legit error(like some job issue) and most probably will not get resolve even with one more repair run), I want to ask what you think for this ? and how can we enable that as well(have one more argument or `databricks_repair_reason_new_settings` is empty, don't repair or something in your mind) ?

I though we already have `repair_run` config?

**Comment by gaurav7261** [comment]
> > @Lee-W should we have one more configurable like user want to repair only in case if `databricks_repair_reason_new_settings` dict key contains the reason, otherwise don't repair at all(as it is a legit error(like some job issue) and most probably will not get resolve even with one more repair run), I want to ask what you think for this ? and how can we enable that as well(have one more argument or `databricks_repair_reason_new_settings` is empty, don't repair or something in your mind) ?
> 
> I though we already have `repair_run` config?

repair_run is just a bool now

**Comment by gaurav7261** [comment]
@Lee-W  I'm thinking for an option not to trigger repair if reason matches nothing in the dict `databricks_repair_reason_new_settings` , what's your thought on this?

**Comment by Lee-W** [comment]
> @Lee-W I'm thinking for an option not to trigger repair if reason matches nothing in the dict `databricks_repair_reason_new_settings` , what's your thought on this?

I don't think we need a new option. If the reason is provided and nothing is matched, then we should not repair. This should be the default behavior. If the user wants to repair at all cases, than the reason dict should not be passed

**Comment by gaurav7261** [comment]
thanks @Lee-W , making the changes, Yes, I also agree with same default behaviour.

**Comment by gaurav7261** [comment]
@Lee-W changes are done, can you please review

**Review by Lee-W — APPROVED** [review]

**Comment by Lee-W** [comment]
LGTM 👍

**Comment by gaurav7261** [comment]
@Lee-W currently we are checking in root state_message only, should we also check `error` key in every fail task array as well ?

**Comment by Lee-W** [comment]
> @Lee-W currently we are checking in root state_message only, should we also check `error` key in every fail task array as well ?

Could you provide an example to illustrate this? Thanks!

**Comment by gaurav7261** [comment]
For example, sometime in case of custom exception, when user want to retry with another job config, in root state_message we get only below one, and main exception in insider array 
```Job run failed with terminal state: {'life_cycle_state': 'INTERNAL_ERROR', 'result_state': 'FAILED', 'state_message': 'Task t1 failed with message: Workload failed, see run output for details. This caused all downstream tasks to get skipped.'} and with the errors [{'task_key': 't1', 'run_id': 293420302889009, 'error': '<span class=\'ansi-red-fg\'>APIError</span>: <!DOCTYPE html>\n<html lang=en>\n  <meta charset=utf-8>\n  <meta name=viewport content="initial-scale=1, minimum-scale=1, width=device-width">\n  <title>Error 502 (Server Error)!!1</title>\n  <style>\n    {margin:0;padding:0}html,code{font:15px/22px arial,sans-serif}html{background:#fff;color:#222;padding:15px}body{margin:7% auto 0;max-width:390px;min-height:180px;padding:30px 0 15px} > body{background:url(//www.google.com/images/errors/robot.png) 100% 5px no-repeat;padding-right:205px}p{margin:11px 0 22px;overflow:hidden}ins{color:#777;text-decoration:none}a img{border:0}@media screen and (max-width:772px){body{background:none;margin-top:0;max-width:none;padding-right:0}}#logo{background:url(//www.google.com/images/branding/googlelogo/1x/googlelogo_color_150x54dp.png) no-repeat;margin-left:-5px}@media only screen and (min-resolution:192dpi){#logo{background:url(//www.google.com/images/branding/googlelogo/2x/googlelogo_color_150x54dp.png) no-repeat 0% 0%/100% 100%;-moz-border-image:url(//www.google.com/images/branding/googlelogo/2x/googlelogo_color_150x54dp.png) 0}}@media only screen and (-webkit-min-device-pixel-ratio:2){#logo{background:url(//www.google.com/images/branding/googlelogo/2x/googlelogo_color_150x54dp.png) no-repeat;-webkit-background-size:100% 100%}}#logo{display:inline-block;height:54px;width:150px}\n  </style>\n  <a href=//www.google.com/><span id=logo aria-label=Google></span></a>\n  <p><b>502.</b> <ins>That’s an error.</ins>\n  <p>The server encountered a temporary error and could not complete your request.<p>Please try again in 30 seconds.  <ins>That’s all we know.</ins>\n'}]```

**Comment by Lee-W** [comment]
It looks like the error is part of the `state_message`. the user can still match it in current implementation? or is there anything I missed

**Comment by gaurav7261** [comment]
ya but not always, sometime root `state_message` = `Task t1, t2... failed with message: Workload failed, see run output for details. This caused all downstream tasks to get skipped.`, nothing more information, in that case can we also allow matching with `error` key inside array of tasks? what you think @Lee-W , what I have seen in case of AWS side failures, we get root state_message correct, but when user throw custom exception or some databricks errors, state_message is same line `Task failed: Workload failed, see run output for details...`, nothing about main exception string

**Comment by Lee-W** [comment]
if that's the case, it makes sense to have one

**Comment by gaurav7261** [comment]
cool, will create a new pr for the same. Thanks

**Pull request merged**